### DISCUSSION TITLE
Server running out of memory

### DISCUSSION THREAD

**Discussion by dorinclisu** [bluenviron/mediamtx#793] [Questions]
## Which version are you using?

v0.16.2

## Which operating system are you using?
OS
- [x] Linux
- [ ] Windows
- [ ] macOS

Architecture
- [x] amd64
- [ ] arm64
- [ ] arm7
- [ ] arm6

## Describe the issue
I suspect my instance is crashing due to running out of RAM. There is no logs out of the ordinary, it just hangs with cpu usage maxed out until killed. Last log line:
```
2021/06/21 08:54:36 I [2/71] [RTSP] [conn A.B.C.D:3661] closed
```

Is this the expected behavior when the memory runs out? If so, can something be made to detect memory allocation failure and log a suggestive message?

**Comment by aler9** [comment]
Hello,
Memory allocation failures can't be detected in Go. Anyway, this could be a bug instead of a simple exhaustion of RAM, are you certain that the root cause is the RAM?

**Comment by dorinclisu** [comment]
I cannot be 100% sure it is RAM, but there is no other clue in the logs and since then I increased the available RAM and the issue disappeared.

In any case this is problematic compared to other programs I am managing, because in general if a process is running out of memory the container is getting killed with 137 return code, which makes it clear of the cause even without failure detection in the process code.

But here the container was not getting killed, and once the freeze was noticeable, the RAM had been freed already. This tells me that the server fails before the container reaches its allocated memory limit. I don't know how Go works, could it be that the stack and the heap clash due to memory fragmentation, and this manifests only when the free memory is relatively small?

**Comment by aler9** [comment]
I experienced RAM exhaustion recently and i can confirm that the server gets killed with code 137. To make sure of this, you can also run the server inside a container, using the `-m` flag. In this way, the server will be killed before the RAM gets filled up:

```
docker run --rm -it -m 4G ... aler9/rtsp-simple-server
```

[a comment by an automated account is not shown]
